### PR TITLE
Start working on the cache for Conan 2.0

### PR DESCRIPTION
**Pull request by czoido**
Changelog: Feature: Implement new multi-revision cache.
Docs: missing (2.0)

Starting taking some ideas from: https://github.com/conan-io/conan/pull/8510
This PR attempts to make a new cache using a simple sqlite database.

Some of the features for the new Conan Cache in 2.0 are:
- Support for multiple revisions
- Concurrent cache. Use of locks: multiprocess required, also could be multithread
- Multiple levels cache: Read-Only and Write cache

This PR will be mainly focused in the support for multiple package and recipe revisions...

Will close some parts of: https://github.com/conan-io/conan/issues/9076

### CODE REVIEW

File: conans/client/installer.py [hunk 1]
```diff
@@ -491,49 +503,54 @@ def _handle_node_editable(self, node, profile_host, profile_build, graph_lock):
...
-            assert os.path.isdir(package_folder), ("Package '%s' folder must exist: %s\n"
-                                                   % (str(pref), package_folder))
-            # Call the info method
-            self._call_package_info(conanfile, package_folder, ref=pref.ref, is_editable=False)
-            self._recorder.package_cpp_info(pref, conanfile.cpp_info)
-
-    def _build_package(self, node, output, remotes):
+        pkg_layout = pkg_layout or self._cache.pkg_layout(pref)
+        # TODO: cache2.0 Check with new locks
+        # with layout.package_lock(pref):
```
Comment (memsharded) [inline conans/client/installer.py]: Maybe leave an empty, no-op ``package_lock()`` method in the ``layout``. That would highly reduce this diff, and help understand changes.

Then we can do later pure refactors that would remove those things, very easily.

File: conans/client/installer.py [hunk 2]
```diff
@@ -67,9 +68,10 @@ def _get_build_folder(self, conanfile, package_layout, pref, recorder):
         # Build folder can use a different package_ID if build_id() is defined.
         # This function decides if the build folder should be re-used (not build again)
         # and returns the build folder
-        new_id = build_id(conanfile)
+        # TODO: cache2.0 FIX: we are not taking intto account the conanfile build_id
```
Comment (memsharded) [inline conans/client/installer.py]: This seems challenging for the model, maybe one of the first things we want to check if it fits.

File: conan/cache/cache_database.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+from .conan_reference import ConanReference
+from .db.references import ReferencesDbTable
+
+CONNECTION_TIMEOUT_SECONDS = 1  # Time a connection will wait when the database is locked
+
+
+class CacheDatabase:
+    """ Abstracts the operations with the database and ensures they run sequentially """
+    _references = ReferencesDbTable()
```
Comment (memsharded) [inline conan/cache/cache_database.py]: No need to make this a class attribute (singleton-like)

File: conan/cache/cache_database.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+    timeout = CONNECTION_TIMEOUT_SECONDS
+
+    def __init__(self, filename):
+        self._filename = filename
+
+    @contextmanager
+    def connect(self):
+        conn = sqlite3.connect(self._filename, isolation_level=None, timeout=self.timeout)
```
Comment (memsharded) [inline conan/cache/cache_database.py]: This might be quite expensive, to open a connection every time, for every query. 
Why not keeping an open connection? That is a typical access patterns from apps.

File: conan/cache/cache_database.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        with self.connect() as conn:
+            output.write(f"\nReferencesDbTable (table '{self._references.table_name}'):\n")
+            self._references.dump(conn, output)
+
+    def update_reference(self, old_ref: ConanReference, new_ref: ConanReference = None,
+                         new_path=None, new_remote=None):
+        with self.connect() as conn:
+            try:
+                self._references.update(conn, old_ref, new_ref, new_path, new_remote)
+            except sqlite3.IntegrityError:
```
Comment (memsharded) [inline conan/cache/cache_database.py]: Why? Concerning this exception capture here.

File: conan/cache/cache_database.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+    def get_or_create_reference(self, path, ref: ConanReference):
+        """ Returns the path for the given reference. If the reference doesn't exist in the
+            database, it will create the entry for the reference using the path given as argument.
+        """
+        with self.connect() as conn:
+            try:
+                ref_data = self._references.get(conn, ref)
+                return ref_data["path"], False
+            except ReferencesDbTable.DoesNotExist:
```
Comment (memsharded) [inline conan/cache/cache_database.py]: I'd create the table on connection/opening if it doesn't exist, not here.

File: conans/client/cmd/export.py [hunk 7]
```diff
@@ -33,9 +34,7 @@ class AliasConanfile(ConanFile):
     manifest.save(folder=package_layout.export())
 
     # Create the metadata for the alias
-    _update_revision_in_metadata(package_layout=package_layout,
-                                 output=output, path=None, manifest=manifest,
-                                 revision_mode=revision_mode)
+    calc_revision(output=output, path=None, manifest=manifest, revision_mode=revision_mode)
```
Comment (memsharded) [inline conans/client/cmd/export.py]: This seems have no effect?

File: conans/client/cmd/export_pkg.py [hunk 8]
```diff
@@ -32,29 +32,30 @@ def export_pkg(app, recorder, full_ref, source_folder, build_folder, package_fol
     package_id = pkg_node.package_id
     output.info("Packaging to %s" % package_id)
     pref = PackageReference(ref, package_id)
-    layout = cache.package_layout(ref, short_paths=conanfile.short_paths)
-
-    if layout.package_id_exists(package_id) and not force:
+    pkg_ids = cache.get_package_ids(ref)
+    existing_id = [pkg_id for pkg_id in pkg_ids if package_id in pkg_id]
```
Comment (memsharded) [inline conans/client/cmd/export_pkg.py]: Weird, probably just ``pkg_id in pkg_ids``?

File: conans/client/downloaders/file_downloader.py [hunk 9]
```diff
@@ -70,7 +70,7 @@ def _download_file(self, url, auth, headers, file_path, try_resume=False):
             range_start = 0
 
         try:
-            response = self._requester.get(url, stream=True, verify=self._verify_ssl, auth=auth,
+            response = self._requester.get(url, stream=True, verify=False, auth=auth,
```
Comment (memsharded) [inline conans/client/downloaders/file_downloader.py]: How is this related to the PR?

File: conan/cache/cache.py [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+    def _full_path(self, relative_path):
+        path = os.path.realpath(os.path.join(self._base_folder, relative_path))
+        return path
+
+    @property
+    def base_folder(self):
+        return self._base_folder
+
+    @staticmethod
+    def get_or_create_reference_path(ref: ConanReference):
```
Comment (memsharded) [inline conan/cache/cache.py]: Make it class private, not used elsewhere

File: conan/cache/cache.py [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+        """ Returns a folder for a Conan-Reference, it's deterministic if revision is known """
+        if ref.rrev:
+            return md5(ref.full_reference)
+        else:
+            return str(uuid.uuid4())
+
+    @staticmethod
+    def get_or_create_package_path(ref: ConanReference):
```
Comment (memsharded) [inline conan/cache/cache.py]: Make it class private

File: conan/cache/cache.py [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+        path = self.get_or_create_reference_path(ref)
+
+        if not ref.rrev:
+            ref = ConanReference(ref.name, ref.version, ref.user, ref.channel, path,
+                                 ref.pkgid, ref.prev)
+
+        reference_path, created = self.db.get_or_create_reference(path, ref)
+        self._create_path(reference_path, remove_contents=created)
+
+        from conan.cache.conan_reference_layout import ReferenceLayout
```
Comment (memsharded) [inline conan/cache/cache.py]: Really necessary? Circular import otherwise?

File: conan/cache/cache.py [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+        return ReferenceLayout(ref, cache=self, base_folder=reference_path)
+
+    def get_or_create_package_layout(self, pref: ConanReference):
+        package_path = self.get_or_create_package_path(pref)
+
+        # Assign a random (uuid4) revision if not set
+        # if the package revision is not calculated yet, assign the uuid of the path as prev
+        # TODO: cache2.0: fix this in the future
+        rrev = pref.rrev or package_path
+        prev = pref.prev or package_path
```
Comment (memsharded) [inline conan/cache/cache.py]: This is really confusing, constructing a ConanReference with a path inside the revisions fields

File: conan/cache/conan_reference_layout.py [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+        new_path = self._cache._move_rrev(old_ref, self._ref)
+        if new_path:
+            self._base_folder = new_path
+
+    @property
+    def base_folder(self):
+        return os.path.join(self._cache.base_folder, self._base_folder)
+
+    def build(self):
+        assert self._ref.pkgid, "Must be a reference of a package"
```
Comment (memsharded) [inline conan/cache/conan_reference_layout.py]: I am not fully convinced about this, it might be better to split the class (same creation interface, transparent instantiation)

File: conan/cache/db/table.py [hunk 15]
```diff
@@ -0,0 +1,51 @@
+from collections import namedtuple
+from io import StringIO
+from typing import Tuple, List, Optional
+
+
+class BaseDbTable:
+    table_name: str = None
+    columns_description: List[Tuple[str, type]] = None
```
Comment (memsharded) [inline conan/cache/db/table.py:8]: Seems the type doesn't match, as the columns can be a tuple of ``(str, type, boolean, ....)``

File: conan/cache/cache_database.py (hunk 6 above)
Comment (czoido) [reply conan/cache/cache_database.py]: Table creation is made when the cache is initialised, maybe the exception name is confusing, I'll change that to _ReferenceDoesNotExist_

**Review by memsharded — APPROVED** [review]
